File: backend/runner/pubsub/testdata/java/publisher/src/main/java/xyz/block/ftl/java/test/publisher/Publisher.java
package xyz.block.ftl.java.test.publisher;

import io.quarkus.logging.Log;
import xyz.block.ftl.Export;
import xyz.block.ftl.FromOffset;
import xyz.block.ftl.Subscription;
import xyz.block.ftl.Topic;
import xyz.block.ftl.TopicPartitionMapper;
import xyz.block.ftl.Verb;
import xyz.block.ftl.WriteableTopic;

class PartitionMapper implements TopicPartitionMapper<PubSubEvent> {
    public String getPartitionKey(PubSubEvent event) {
        return event.getTime().toString();
    }
}

public class Publisher {

    @Export
    @Topic("testTopic")
    interface TestTopic extends WriteableTopic<PubSubEvent, PartitionMapper> {

    }

    @Topic("localTopic")
    interface LocalTopic extends WriteableTopic<PubSubEvent, PartitionMapper> {

    }

    @Export
    @Topic("topic2")
    interface Topic2 extends WriteableTopic<PubSubEvent, PartitionMapper> {

    }

    @Verb
    void publishTen(LocalTopic testTopic) throws Exception {
        for (var i = 0; i < 10; ++i) {
            var t = java.time.ZonedDateTime.now();
            Log.infof("Publishing %s", t);
            testTopic.publish(new PubSubEvent().setTime(t));
        }
    }

    @Verb
    void publishOne(TestTopic testTopic) throws Exception {
        var t = java.time.ZonedDateTime.now();
        Log.infof("Publishing %s", t);
        testTopic.publish(new PubSubEvent().setTime(t));
    }

    @Verb
    void publishOneToTopic2(HaystackRequest req, Topic2 topic2) throws Exception {
        var t = java.time.ZonedDateTime.now();
        Log.infof("Publishing %s", t);
        topic2.publish(new PubSubEvent().setTime(t).setHaystack(req.getHaystack()));
    }

    @Subscription(topic = LocalTopic.class, from = FromOffset.LATEST)
    public void local(TestTopic testTopic, PubSubEvent event) {
        testTopic.publish(event);
    }
}

File: backend/runner/pubsub/testdata/kotlin/publisher/src/main/kotlin/xyz/block/ftl/java/test/publisher/Publisher.kt
package xyz.block.ftl.java.test.publisher

import io.quarkus.logging.Log
import xyz.block.ftl.*
import java.time.ZonedDateTime

class PartitionMapper : TopicPartitionMapper<PubSubEvent> {
    override fun getPartitionKey(event: PubSubEvent): String {
        return event.time.toString()
    }
}

class Publisher {
    @Export
    @Topic("testTopic")
    interface TestTopic : WriteableTopic<PubSubEvent, PartitionMapper>

    @Topic("localTopic")
    interface LocalTopic : WriteableTopic<PubSubEvent, PartitionMapper>

    @Export
    @Topic("topic2")
    interface Topic2 : WriteableTopic<PubSubEvent, PartitionMapper>

    @Verb
    @Throws(Exception::class)
    fun publishTen(testTopic: TestTopic) {
        for (i in 0..9) {
            val t = ZonedDateTime.now()
            Log.infof("Publishing to testTopic: %s", t)
            testTopic.publish(PubSubEvent(t))
        }
    }

    @Verb
    @Throws(Exception::class)
    fun publishTenLocal(localTopic: LocalTopic) {
        for (i in 0..9) {
            val t = ZonedDateTime.now()
            Log.infof("Publishing to localTopic: %s", t)
            localTopic.publish(PubSubEvent(t))
        }
    }

    @Verb
    @Throws(Exception::class)
    fun publishOne(testTopic: TestTopic) {
        val t = ZonedDateTime.now()
        Log.infof("Publishing %s", t)
        testTopic.publish(PubSubEvent(t))
    }

    @Verb
    @Throws(Exception::class)
    fun publishOneToTopic2(topic2: Topic2) {
        val t = ZonedDateTime.now()
        Log.infof("Publishing %s", t)
        topic2.publish(PubSubEvent(t))
    }

    @Subscription(topic = LocalTopic::class, from = FromOffset.LATEST)
    fun local(event: PubSubEvent) {
        Log.infof("Consuing from local %s", event.time)
    }
}
